article_store/articles/models.py

```diff
@@ -0,0 +1,84 @@
+import re
+import uuid
+
+from django.core.validators import RegexValidator
+from django.db import models
+
+
+PREVIEW = 'preview'
+PUBLISHED = 'published'
+READY = 'ready'
+
+ARTICLE_STATUSES = (
+    (PREVIEW, 'preview'),
+    (PUBLISHED, 'published'),
+    (READY, 'ready'),
+)
+
+ARTICLE_ID_FORMAT = r'^[A-Za-z0-9\-._]+$'
+
+
+class Article(models.Model):
+    id = models.CharField(primary_key=True, max_length=250,
+                          validators=[RegexValidator(regex=ARTICLE_ID_FORMAT)])
+
+    article_id_format = ARTICLE_ID_FORMAT
+    default_version = 0
+
+    def __str__(self):
+        return f'{self.id}'
+
+    @property
+    def latest_version(self) -> int:
+        """Finds the latest associated `ArticleVersion.version`.
+
+        :return: int
+        """
+        versions = self.versions.all()
+
+        if versions:
+            return max([version.version for version in versions ])
+
+        return self.default_version
+
+    @property
+    def next_version(self) -> int:
+        """Finds the next version from the associated `ArticleVersion.version`.
+
+        :return: int
+        """
+        return self.latest_version + 1
+
+    @classmethod
+    def id_is_valid(cls, article_id) -> bool:
+        """Check if a `str` value is valid against `article_id_format`
+
+        :param article_id: str
+        :return: bool
+        """
+        return True if re.match(cls.article_id_format, article_id) else False
+
+
+class ArticleVersion(models.Model):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
```

Comment:
UUID for `ArticleVersion` may be an overkill because there is a strict order of versions and they must be generated over a single-master, linearizable database (`postgres`).

Since `article`+`version` is a composite key I agree a generated primary key makes things easier especially for the ORM. Could just be an incremental id? It's never going to be exposed to the API.

Reply:
Will change to incremental id for simplicity then.

Reply:
It's not going to allow a nice migration from `uuid` to `int`:

`cannot cast type uuid to integer`

I will reset the migrations to wipe the `uuid` column, will just mean you will have to start clean.